datasource/ethereum/src/block_ingestor.rs

```diff
@@ -0,0 +1,178 @@
+use failure::Error;
+use std::fmt::Debug;
+use std::sync::Mutex;
+use std::time::Duration;
+use std::time::Instant;
+use web3::api::Web3;
+use web3::transports::batch::Batch;
+use web3::types::Block;
+use web3::types::BlockId;
+use web3::types::BlockNumber;
+use web3::types::H256;
+use web3::types::Transaction;
+use web3::BatchTransport;
+use web3::Transport;
+
+use graph::prelude::*;
+
+pub struct BlockIngestor<S, T>
+where
+    S: BlockStore + Send + 'static,
+    T: BatchTransport + Send + Sync + Debug + Clone + 'static,
+    <T as Transport>::Out: Send,
+    <T as BatchTransport>::Batch: Send,
+{
+    store: Arc<Mutex<S>>,
+    network_name: String,
+    web3_transport: T,
+    ancestor_count: u64,
+    logger: slog::Logger,
+    polling_interval: Duration,
+}
+
+impl<S, T> BlockIngestor<S, T>
+where
+    S: BlockStore + Send + 'static,
+    T: BatchTransport + Send + Sync + Debug + Clone + 'static,
+    <T as Transport>::Out: Send,
+    <T as BatchTransport>::Batch: Send,
+{
+    pub fn new(
+        store: Arc<Mutex<S>>,
+        network_name: String,
+        web3_transport: T,
+        ancestor_count: u64,
+        logger: slog::Logger,
+        polling_interval: Duration,
+    ) -> Result<BlockIngestor<S, T>, Error> {
+        // Add a head block pointer for this network name if one does not already exist
+        store.lock().unwrap().add_network_if_missing(&network_name)?;
+
+        Ok(BlockIngestor {
+            store,
+            network_name,
+            web3_transport,
+            ancestor_count,
+            logger: logger.new(o!("component" => "BlockIngestor")),
+            polling_interval,
+        })
+    }
+
+    pub fn into_polling_stream(self) -> impl Future<Item = (), Error = ()> {
+        // Currently, there is no way to stop block ingestion, so just leak self
+        let static_self: &'static _ = Box::leak(Box::new(self));
+
+        // Create stream that emits at polling interval
+        tokio::timer::Interval::new(Instant::now(), static_self.polling_interval)
+            .map_err(move |e| {
+                error!(static_self.logger, "timer::Interval failed: {:?}", e);
+            })
+            .for_each(move |_| {
+                // Attempt to poll
+                static_self.do_poll().then(move |result| {
```

Comment:
This `then` can be refactored as a `map_err`.

Reply:
Hmm does that work? I want to change Err into Ok here to prevent polling from stopping

Reply:
Ah yes, nevermind then.